File: frontend/src/api/processes.ts
import { apiRequest } from "@/api/client";
import { paginatedSchema } from "@/schemas/common";
import { processSchema } from "@/schemas/process";
import type { CreateProcessValues, ProcessListQuery, UpdateProcessValues } from "@/schemas/process";
import type { Paginated, Process } from "@/types/domain";

const listSchema = paginatedSchema(processSchema);

function toQueryString(query: ProcessListQuery): string {
  const params = new URLSearchParams();
  if (query.status) params.set("status", query.status);
  if (query.category) params.set("category", query.category);
  if (query.q) params.set("q", query.q);
  if (query.sort) params.set("sort", query.sort);
  if (query.page) params.set("page", String(query.page));
  if (query.page_size) params.set("page_size", String(query.page_size));
  const s = params.toString();
  return s ? `?${s}` : "";
}

export function listProcesses(query: ProcessListQuery = {}): Promise<Paginated<Process>> {
  return apiRequest(`/processes${toQueryString(query)}`, { schema: listSchema });
}

export function getProcess(id: string): Promise<Process> {
  return apiRequest(`/processes/${id}`, { schema: processSchema });
}

export function createProcess(values: CreateProcessValues): Promise<Process> {
  return apiRequest("/processes", { method: "POST", body: values, schema: processSchema });
}

export function updateProcess(id: string, values: UpdateProcessValues): Promise<Process> {
  return apiRequest(`/processes/${id}`, { method: "PATCH", body: values, schema: processSchema });
}

export function deleteProcess(id: string): Promise<void> {
  return apiRequest(`/processes/${id}`, { method: "DELETE" });
}

File: backend/app/api/v1/processes.py
from __future__ import annotations

from fastapi import APIRouter, Depends, Query
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.models.task import Task
from app.models.user import User
from app.schemas.common import Page
from app.schemas.process import ProcessCreate, ProcessRead, ProcessUpdate
from app.schemas.task import TaskCreate, TaskRead
from app.security import get_current_user
from app.services import process_service

router = APIRouter(prefix="/processes", tags=["processes"])


@router.get("", response_model=Page[ProcessRead])
async def list_processes(
    status: str | None = None,
    category: str | None = None,
    q: str | None = None,
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> Page[ProcessRead]:
    rows, total = await process_service.list_processes(
        db, current_user.id, status=status, category=category, q=q, page=page, page_size=page_size
    )
    return Page[ProcessRead](
        items=[ProcessRead.model_validate(r) for r in rows], total=total, page=page, page_size=page_size
    )


@router.post("", response_model=ProcessRead, status_code=201)
async def create_process(
    payload: ProcessCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> ProcessRead:
    process = await process_service.create_process(db, current_user.id, payload)
    return ProcessRead.model_validate(process)


@router.get("/{process_id}", response_model=ProcessRead)
async def get_process(
    process_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> ProcessRead:
    process = await process_service.get_process_or_404(db, current_user.id, process_id)
    return ProcessRead.model_validate(process)


@router.patch("/{process_id}", response_model=ProcessRead)
async def update_process(
    process_id: str,
    payload: ProcessUpdate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> ProcessRead:
    process = await process_service.update_process(db, current_user.id, process_id, payload)
    return ProcessRead.model_validate(process)


@router.delete("/{process_id}", status_code=204, response_model=None)
async def delete_process(
    process_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> None:
    await process_service.delete_process(db, current_user.id, process_id)


@router.get("/{process_id}/tasks", response_model=list[TaskRead])
async def list_process_tasks(
    process_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> list[TaskRead]:
    # Ownership check at the query level: get_process_or_404 filters by user_id, and the task
    # query below is scoped to that already-verified process_id.
    await process_service.get_process_or_404(db, current_user.id, process_id)
    result = await db.execute(
        select(Task).where(Task.process_id == process_id).order_by(Task.order_index)
    )
    return [TaskRead.model_validate(t) for t in result.scalars().all()]


@router.post("/{process_id}/tasks", response_model=TaskRead, status_code=201)
async def create_process_task(
    process_id: str,
    payload: TaskCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> TaskRead:
    await process_service.get_process_or_404(db, current_user.id, process_id)
    task = Task(process_id=process_id, **payload.model_dump())
    db.add(task)
    await db.commit()
    await db.refresh(task)
    return TaskRead.model_validate(task)
